clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AbstractAvroStoreClient.java

```diff
@@ -272,28 +272,36 @@ protected RecordSerializer<K> getKeySerializerForRequest() {
     }
     synchronized (this) {
       try {
+        if (keySerializer != null) {
+          whetherStoreInitTriggeredByRequestFail = false;
```

Comment:
Looks like we init `keySerializer`, `multiGetRequestSerializer`, `computeRequestClientKeySerializer` and `streamingFooterRecordDeserializer` inside `initSerializer()`, so setting this just by checking `keySerializer` might lead to issues? Even before this resiliency PR, the check and return at line 266 might cause issues for multiget/compute use cases. We probably have to add a minor change in `initSerializer` to set these global variables once all these objects are locally setup.

Reply:
Seems that L266-L268 has a similar check but a bit different in its behavior (missing the statement of whetherStoreInitTriggeredByRequestFail = false; ), is it intended or considerations behind it?

Reply:
@lluwm 
The logic is intended, and the goal is to retrieve `keySerializer` and as long as it is good to use, that function can return.
Right now, the logic around initialization is a little messy, and let me re-think about it for a while to clean it up.

Reply:
Thanks for the explanations, Gaojie. If we don't explicitly set whetherStoreInitTriggeredByRequestFail, at here.
```
    synchronized (this) {
      try {
        if (keySerializer != null) {
          whetherStoreInitTriggeredByRequestFail = false;    <---- here
          return keySerializer;
        }
```
asyncStoreInitThread would still set it to false, after getKeySerializerWithRetryWithShortInterval is done correctly, and because of getKeySerializerWithRetryWithShortInterval is set to volatile, its visibility should also be guaranteed.
```
          // Spin up at most one async thread
          asyncStoreInitThread = new Thread(() -> {
            while (true) {
              try {
                getKeySerializerWithRetryWithShortInterval();
                whetherStoreInitTriggeredByRequestFail = false;  <------here
                LOGGER.info("Successfully init store client by async store init thread");
                break;
              } catch (Exception ee) {
```
whetherStoreInitTriggeredByRequestFail = false; in the 1st code block seems not necessary, but I think your current code still working correctly though.

Reply:
@gaojieliu setting `keySerializer` last is a better fix than the one I suggested. Thanks for changing that.